src/lower/mod.rs

```diff
@@ -1407,64 +1412,126 @@ impl ir::AssociatedTyDatum {
-            binders.push(ir::ParameterKind::Ty(()));
-            let ty = ir::Ty::Var(binders.len() - 1);
-
-            ir::ProgramClause {
-                implication: ir::Binders {
-                    binders,
-                    value: ir::ProgramClauseImplication {
-                        consequence: ir::ProjectionEq {
-                            projection: projection.clone(),
-                            ty: ty.clone(),
-                        }.cast(),
-                        conditions: vec![
-                            ir::Normalize {
-                                projection: projection.clone(),
-                                ty,
-                            }.cast()
-                        ],
-                    },
+        clauses.push(ir::ProgramClause {
+            implication: ir::Binders {
+                binders: binders.clone(),
+                value: ir::ProgramClauseImplication {
+                    consequence: ir::ProjectionEq {
+                        projection: projection.clone(),
+                        ty,
+                    }.cast(),
+                    conditions: vec![normalize.clone().cast()],
+                },
+            },
+            });
+
+
+        // We generate a proxy rule for the well-formedness of `T: Foo<Assoc = U>` which really
+        // means two things: `T: Foo` and `Normalize(<T as Foo>::Assoc -> U)`. So we have the
+        // following rule:
+        //
+        //    forall<T> {
+        //        WellFormed(T: Foo<Assoc = U>) :-
+        //            WellFormed(T: Foo), Normalize(<T as Foo>::Assoc -> U)
+        //    }
+        clauses.push(ir::ProgramClause {
+            implication: ir::Binders {
+                binders: binders.clone(),
+                value: ir::ProgramClauseImplication {
+                    consequence: ir::WellFormed::Normalize(normalize.clone()).cast(),
+                    conditions: vec![
+                        normalize.clone().cast(),
+                        ir::WellFormed::TraitRef(trait_ref.clone()).cast()
+                    ],
+                }
+            }
+        });
+
+        // We also have two proxy reverse rules, the first one being:
+        //
+        //    forall<T> {
+        //        FromEnv(T: Foo) :- FromEnv(T: Foo<Assoc = U>)
+        //    }
+        clauses.push(ir::ProgramClause {
+            implication: ir::Binders {
+                binders: binders.clone(),
+                value: ir::ProgramClauseImplication {
+                    consequence: ir::FromEnv::TraitRef(trait_ref).cast(),
+                    conditions: vec![ir::FromEnv::Normalize(normalize.clone()).cast()],
                 },
             }
-        };
+        });
 
-        vec![fallback_clause, normalize_clause]
+        // And the other one being:
+        //
+        //    forall<T> {
+        //        Normalize(<T as Foo>::Assoc -> U) :- FromEnv(T: Foo<Assoc = U>)
+        //    }
```

Comment:
As an aside, and not for this PR, I wonder if we can use macros to make this code less tedious to write.